jsonrpc/debug_endpoint.go

```diff
@@ -667,6 +686,71 @@ func (d *Debug) GetRawReceipts(filter BlockNumberOrHash) (interface{}, error) {
 	)
 }
 
+func (d *Debug) TraceChain(start, end BlockNumber, config *TraceConfig) (interface{}, error) {
+	return d.throttling.AttemptRequest(
+		context.Background(),
+		func() (interface{}, error) {
+			startNum, err := GetNumericBlockNumber(start, d.store)
+			if err != nil {
+				return nil, err
+			}
+
+			endNum, err := GetNumericBlockNumber(end, d.store)
+			if err != nil {
+				return nil, err
+			}
+
+			if startNum >= endNum {
+				return nil, fmt.Errorf("end block (#%d) needs to come after start block (#%d)", endNum, startNum)
+			}
+
+			blocks := int(endNum-startNum) + 1
+			results := make([]BlockTraceResult, blocks)
+			resultCh := make(chan BlockTraceResult, blocks)
+
+			var wg sync.WaitGroup
+
+			for i := 0; i < blocks; i++ {
+				wg.Add(1)
+
+				go func(i int) {
+					defer wg.Done()
+
+					blockNum := startNum + uint64(i)
+					traceResult := BlockTraceResult{Block: blockNum}
+
+					block, ok := d.store.GetBlockByNumber(blockNum, true)
+					if !ok {
+						traceResult.Error = fmt.Errorf("block %d not found", blockNum)
+					} else {
+						traceResult.Hash = block.Hash()
+
+						res, err := d.traceBlock(block, config)
+						if err != nil {
+							traceResult.Error = fmt.Errorf("failed to trace block %d: %w", blockNum, err)
+						} else {
+							traceResult.Traces = res
+						}
+					}
+
+					resultCh <- traceResult
+				}(i)
+			}
+
+			go func() {
+				wg.Wait()
+				close(resultCh)
+			}()
```

Comment:
We don't need this go routine. Just call wg.Wait() to wait for all trace routines to finish and close(resultCh) after for loop (before return).

Reply:
By using a separate goroutine, the main function can continue reading from resultCh as each result is sent, rather than waiting until all goroutines finish. The channel is closed exactly when all results have been gathered, indicated by wg.Wait().

Here it might make sense to add a semaphore to limit CPU usage to the number of cores minus one (leaving one thread free). Does this make sense to you?

Reply:
As Stefan mentioned, we don't need semaphore and waitgroup can also be removed. Consider this implementation, it will work in parallel with only 1 channel and without semaphore, waitgroup and additiona go routine:
			resultCh := make(chan BlockTraceResult, 1)

			for i := 0; i < blocks; i++ {
				go func(i int) {
					blockNum := startNum + uint64(i)
					traceResult := BlockTraceResult{Block: blockNum}

					block, ok := d.store.GetBlockByNumber(blockNum, true)
					if !ok {
						traceResult.Error = fmt.Errorf("block %d not found", blockNum)
					} else {
						traceResult.Hash = block.Hash()

						res, err := d.traceBlock(block, config)
						if err != nil {
							traceResult.Error = fmt.Errorf("failed to trace block %d: %w", blockNum, err)
						} else {
							traceResult.Traces = res
						}
					}

					resultCh <- traceResult
				}(i)
			}

			traceResults := 0

			for {
				select {
				case traceResult := <-resultCh:
					results[traceResult.Block-startNum] = traceResult

					traceResults++
					if traceResults == blocks {
						return results, nil
					}
				}
			}